### PR TITLE
[A11y] - Accessibility in Ask Forms

### PR DESCRIPTION
**Pull request by okbel**
# Accessibility
### Things I've learned and Resources

An amazing course by [Marcy Sutton](http://twitter.com/marcysutton) 

> https://egghead.io/courses/start-building-accessible-web-applications-today

The usage of labelledby

> https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-labelledby_attribute

Relying on the title attribute

> https://www.paciellogroup.com/blog/2012/01/html5-accessibility-chops-title-attribute-use-and-abuse/

A guide to use good practices to make your sites accessible

> http://webaim.org/techniques/forms/controls

Another great article for building forms

> https://24ways.org/2015/the-accessibility-mindset/

How to structure forms

> https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Forms/How_to_structure_an_HTML_form

W3C Specification on A11y

> https://www.w3.org/TR/WCAG20/ 
> Tutorials from the W3C to make accessble sites
> https://www.w3.org/WAI/tutorials/forms/
> https://www.w3.org/WAI/intro/wcag.php
> https://www.w3.org/WAI/tutorials/forms/labels/

Tools to test your site

> http://wave.webaim.org/report#/https://coralproject.net/contact.html

Who to follow to know more about a11y

> Talita Pagani [@talitapagani](http://twitter.com/talitapagani) 
> Marcy Sutton [@marcysutton](http://twitter.com/marcysutton) 
## Whats new with this PR?
- Accesibility added in Ask Forms
## How to test this PR?
- Run Cay and Elkhorn
- Create a new Form
- Add all the fields you can
- Check the source code and look the for attr and id usage with labels
- Use a Screen Reader or fire VoiceOver in Mac 

Note: I couldn't test the standalone form due to a auth issue 

------------------------------ EDIT: 

Dealing with this Chrome issue lead my to find a lot of how a user interacts with the accesibility tools. 

I found this article http://webaim.org/techniques/keyboard/ that tells you how a user uses the keyboard when he's navigating a Website.
Basically, the way to navigate Radio Butons is with `<Arrows Keys>` or pressing `<Control> + <Option> + <Arrow Key>`

screen shot 2016-10-19 at 12 18 11 pm

To test: 
- Go to this form that I just created with this branch - https://s3.amazonaws.com/nyt-stage/580004390cefe54d86e2c731.html
- Turn on VoiceOver `<Command> + <f5>`
- Navigate with `<Tab>` 
- When you get to the Radiobuttons you will get a message like this

> screen shot 2016-10-19 at 12 27 18 pm
- Wait 3 seconds and this message should appear: (People that already use the software already know this, if not you wait and you get this hint)

> screen shot 2016-10-19 at 12 22 57 pm
- Navigate using `<Arrows Keys>` or pressing `<Control> + <Option> + <Arrow Key>`
# DEMO TIME :D!

a11yoptions

### CODE REVIEW

File: src/components/AskFieldWrapper.js [hunk 1]
```diff
@@ -77,15 +77,27 @@ class AskFieldWrapper extends Component {
...
           {
             this.props.type === 'field'
-            ? <div>
-              <h3 onMouseDown={this.onMouseDown.bind(this)} onKeyDown={this.onKeyDown.bind(this)} title={'Field number ' + this.props.fieldNumber} tabindex='0' style={this.getTitleStyles()}>
+            ? <label
+                for={`${component}--${fieldNumber}`}
+                style={styles.label}
+              >
+              <h3
+                onMouseDown={this.onMouseDown.bind(this)}
```
Comment (davidgljay) [inline src/components/AskFieldWrapper.js]: Let's move binding to the constructor rather than rebinding on render (unless there's a reason to make an exception here.)

File: src/components/fields/DateField.js [hunk 2]
```diff
@@ -136,39 +136,61 @@ class DateField extends AskField {
...
-          <input
-            ref={el => this._year = el}
-            type="number"
-            step="1"
-            placeholder="YYYY"
-            value={ this.state.year }
-            onChange={ this.onDateChange }
-            style={ this.getDateInputStyles('year') } />
+          <fieldset id={fieldIdentifier}>
+            <label for={`${fieldIdentifier}__month`} hidden>Month</label>
```
Comment (davidgljay) [inline src/components/fields/DateField.js]: Is the intention for this to be visible to a screen reader only? It looks like this is the same as `display: hidden;`, which would also render the label invisible to screen readers. I tested [here](http://www.w3schools.com/tags/tryit.asp?filename=tryhtml5_global_hidden) to confirm.

Maybe use this instead?

'.visuallyhidden { 
  position: absolute; 
  overflow: hidden; 
  clip: rect(0 0 0 0); 
  height: 1px; width: 1px; 
  margin: -1px; padding: 0; border: 0; 
}'

From https://snook.ca/archives/html_and_css/hiding-content-for-accessibility

File: src/components/AskFieldWrapper.js (hunk 1 above)
Comment (okbel) [reply src/components/AskFieldWrapper.js]: Changing, I missed that one!

File: src/components/fields/DateField.js (hunk 2 above)
Comment (okbel) [reply src/components/fields/DateField.js]: I remember that from the HTML5 boilerplate. Adding! Thanks!

**Comment by davidgljay** [comment]
Forms seem to work with a screen reader except for multiple choice forms:

image

The header is read, but the individual options are skipped over, possibly because they're wrapped by a div.

File: src/components/fields/DateField.js [hunk 3]
```diff
@@ -136,39 +136,61 @@ class DateField extends AskField {
...
+            />
+          </fieldset>
         </div>
-        <div style={ styles.calendarButton } ref={el => this._calendarGroup = el} data-wrap data-clickOpens="false">
+        <div
+          role="button"
+          tabIndex="0"
+          style={ styles.calendarButton }
+          ref={el => this._calendarGroup = el}
+          data-wrap data-clickOpens="false"
```
Comment (davidgljay) [inline src/components/fields/DateField.js:191]: It might be good to include a description of this button for screen readers and make it accessible on keydown:

`<span style={styles.visuallyHidden}>Calendar Widget</span>`

Trick is, the default calendar widget isn't screen reader accessible (it just reads off a long list of dates), and fixing that seems like a broader task. I wonder if we should just make it aria-hidden for now, since there's another more convenient way to enter dates with a keyboard.

**Comment by okbel** [comment]
@davidgljay I'm using Safari for VoiceOver since Safari has better accessibility support than Chrome does. Multiple choice works well and as expected on my end. 

About the Calendar button: I agree. Theres no point making it clickable. Also, the screen reader goes a bit crazy with that calendar widget. We might replace it in the future.

**Comment by okbel** [comment]
Dealing with this Chrome issue lead my to find a lot of how a user interacts with the accesibility tools. 

I found this article http://webaim.org/techniques/keyboard/ that tells you how a user uses the keyboard when he's navigating a Website.
Basically, the way to navigate Radio Butons is with `<Arrows Keys>` or pressing `<Control> + <Option> + <Arrow Key>`

screen shot 2016-10-19 at 12 18 11 pm

To test: 
- Go to this form that I just created with this branch - https://s3.amazonaws.com/nyt-stage/580004390cefe54d86e2c731.html
- Turn on VoiceOver `<Command> + <f5>`
- Navigate with `<Tab>` 
- When you get to the Radiobuttons you will get a message like this

> screen shot 2016-10-19 at 12 27 18 pm
- Wait 3 seconds and this message should appear: (People that already use the software already know this, if not you wait and you get this hint)

> screen shot 2016-10-19 at 12 22 57 pm
- Navigate using `<Arrows Keys>` or pressing `<Control> + <Option> + <Arrow Key>`

# DEMO TIME :D!

a11yoptions

**Comment by riley** [comment]
this is crazy awesome. still working through all the resources you provided here 💯 💯 💯
